sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1098,29 +1100,37 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
   }
 
   private def writeStructToStringBuilder(
-      st: Seq[DataType],
+      st: Seq[StructField],
       row: ExprValue,
       buffer: ExprValue,
       ctx: CodegenContext): Block = {
-    val structToStringCode = st.zipWithIndex.map { case (ft, i) =>
-      val fieldToStringCode = castToStringCode(ft, ctx)
-      val field = ctx.freshVariable("field", ft)
-      val fieldStr = ctx.freshVariable("fieldStr", StringType)
-      val javaType = JavaCode.javaType(ft)
-      code"""
-         |${if (i != 0) code"""$buffer.append(",");""" else EmptyBlock}
-         |if ($row.isNullAt($i)) {
```

Comment:
When the actual value is null, for primitive type field, `row.isNullAt(i)` return ture, but `row.getXXX` return a default value.

For exmaple:

```scala
val r = new org.apache.spark.sql.catalyst.expressions.GenericInternalRow(Array(1, null))
println(r.getInt(0))   // 1
println(r.getInt(1))   // 0
println(r.isNullAt(1)) // true
```

so we cann't only check `row.isNullAt(i)` here, we need to do the same logical like  `BoundReference.doGenCode()`, add nullable check.